packages/flutter_tools/lib/src/cache.dart

```
@@ -1487,3 +1388,134 @@ const List<List<String>> _dartSdks = <List<String>> [
+  final FileSystem _fileSystem;
+  final Directory _tempStorage;
+
+  /// Keep track of the files we've downloaded for this execution so we
+  /// can delete them after completion. We don't delete them right after
+  /// extraction in case [update] is interrupted, so we can restart without
+  /// starting from scratch.
+  @visibleForTesting
+  final List<File> downloadedFiles = <File>[];
+
+  /// Download a zip archive from the given [url] and unzip it to [location].
+  Future<void> downloadZipArchive(
+    String message,
+    Uri url,
+    Directory location,
+  ) {
+    return _downloadArchive(
+      message,
+      url,
+      location,
+      _operatingSystemUtils.unzip,
+    );
+  }
+
+  /// Download a gzipped tarball from the given [url] and unpack it to [location].
+  Future<void> downloadZippedTarball(String message, Uri url, Directory location) {
+    return _downloadArchive(
+      message,
+      url,
+      location,
+      _operatingSystemUtils.unpack,
+    );
+  }
+
+  /// Download an archive from the given [url] and unzip it to [location].
+  Future<void> _downloadArchive(
+    String message,
+    Uri url,
+    Directory location,
+    void Function(File, Directory) extractor,
+  ) async {
+    final String downloadPath = flattenNameSubdirs(url, _fileSystem);
+    final File tempFile = _createDownloadFile(downloadPath);
+    final Status status = _logger.startProgress(message, timeout: const Duration(minutes: 2));
+    int retries = 0;
+
+    while (retries < 2) {
+      try {
+        _ensureExists(tempFile.parent);
+        await _net.fetchUrl(url, destFile: tempFile, maxAttempts: 2);
+      } finally {
+        status.stop();
+      }
+      _ensureExists(location);
+      try {
+        extractor(tempFile, location);
+      } on ProcessException {
+        if (retries >= 1) {
+          rethrow;
+        }
+        retries += 1;
+        if (tempFile.existsSync()) {
+          tempFile.deleteSync();
```

Ignore exceptions thrown when trying to delete the file.